File: sentimental-credit/credit.py
import sys
from cs50 import get_int


def main():
    # Get card as int
    card = get_int("Number: ")

    # Checks card length
    length = len(str(card))
    if not 13 <= length <= 16:
        print("INVALID")
        sys.exit()

    # Check card type
    type = check_type(card, length)
    if type == "INVALID":
        print("INVALID")
        sys.exit()

    # Check card digits sum
    sum = check_sum(card, length) % 10

    # If sum is correct (0) the card is valid
    if sum == 0:
        print(type)
        sys.exit()
    else:
        print("INVALID")
        sys.exit()


def check_type(card, length):
    card = str(card)
    first_digit = card[0]
    first_two = card[:2]

    if ((length == 15) and (first_two == "37" or first_two == "34")):
        return "AMEX"
    elif ((length == 13 or length == 16) and (first_digit == "4")):
        return "VISA"
    elif ((length == 16) and (first_two in ["51", "52", "53", "54", "55"])):
        return "MASTERCARD"
    else:
        return "INVALID"


def check_sum(card, length):
    card = str(card)
    sum = 0
    digit = 0
    # Iterate from the end of the card (length-1), to the beginning (-1 End not inclusive)
    for i in range(length-1, -1, -1):
        # Get digit
        digit = int(card[i])
        # Add digit to the total sum
        if ((length - 1 - i) % 2) == 0:
            sum += digit
        else:
            if ((digit * 2) > 9):
                sum += ((digit * 2) - 9)
            else:
                sum += (digit * 2)
    return sum


if __name__ == "__main__":
    main()

File: credit/credit.c
// Applies luhn algorithm to check if card number is valid
#include <cs50.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int check_length(long card);
string check_type(long card);
int check_sum(long card, int length);

int main(void)
{
    long card = get_long("Number: ");
    int length = check_length(card);
    string type = check_type(card);
    int sum = check_sum(card, length) % 10;
    if (sum == 0 && strcmp(type, "INVALID") != 0)
    {
        printf("%s\n", type);
    }
    else
    {
        printf("INVALID\n");
    }
}

// Get the card number of digits
int check_length(long card)
{
    int length = 0;
    while (card > 0)
    {
        card /= 10;
        length++;
    }
    return length;
}

// Check if card is of correct type based on card number
string check_type(long card)
{
    int length = check_length(card);
    // Get card first digit
    int first_n = (card / (long) pow(10, length - 1));
    // Get card first two digits
    int first_two = (card / (long) pow(10, length - 2));

    if ((length == 15) && (first_two == 37 || first_two == 34))
    {
        return "AMEX";
    }
    else if ((length == 13 || length == 16) && (first_n == 4))
    {
        return "VISA";
    }
    else if ((length == 16) && (first_two == 51 || first_two == 52 || first_two == 53 || first_two == 54 || first_two == 55))
    {
        return "MASTERCARD";
    }
    else
    {
        return "INVALID";
    }
}

// Luhn algorithm to check if card number is valid
int check_sum(long card, int length)
{
    int sum = 0;
    int digit;
    for (int i = 0; i < length; i++)
    {
        // Get a specific digit of the card based on which iteration
        digit = (card / (long) pow(10, i) % 10);
        // Adds the digit to the total sum
        if (i % 2 == 0)
        {
            sum += digit;
        }
        else
        {
            if ((digit * 2) > 9)
            {
                sum += ((digit * 2) - 9);
            }
            else
            {
                sum += (digit * 2);
            }
        }
    }
    return sum;
}
